db/view/view.cc

```diff
@@ -3593,5 +3597,175 @@ std::chrono::microseconds calculate_view_update_throttling_delay(db::view::updat
         return std::chrono::duration_cast<std::chrono::microseconds>(budget);
     }
 }
+
+static logging::logger vbw_logger("view_building_worker");
+
+// Called in the context of a seastar::thread.
+class view_building_worker::consumer : public view_consumer {
+    lw_shared_ptr<replica::table> _base;
+    dht::decorated_key _current_key;
+
+    mutation_reader& _reader;
+    reader_permit _permit;
+    abort_source& _as;
+
+protected:
+    virtual void check_abort_source() override {
+        _as.check();
+    }
+    virtual bool stop_consuming() override {
+        return _as.abort_requested();
+    }
+    virtual bool stop_consuming_end_of_partition() override {
+        return false;
+    }
+    // Returns if consumer should stop processing
+    virtual bool setup_new_partition() override {
+        return false;
+    }
+
+    virtual dht::decorated_key& get_current_key() override {
+        return _current_key;
+    }
+    virtual void set_current_key(dht::decorated_key key) override {
+        _current_key = std::move(key);
+    }
+
+    virtual lw_shared_ptr<replica::table> base() override {
+        return _base;
+    }
+    virtual mutation_reader& reader() override {
+        return _reader;
+    }
+    virtual reader_permit& permit() override {
+        return _permit;
+    }
+
+public:
+    consumer(std::vector<view_ptr> views_to_build, lw_shared_ptr<replica::table> base, mutation_reader& reader, reader_permit permit, shared_ptr<view_update_generator> gen, gc_clock::time_point now, abort_source& as) 
+        : view_consumer(std::move(gen), now, std::move(views_to_build))
+        , _base(base)
+        , _current_key(dht::minimum_token(), partition_key::make_empty())
+        , _reader(reader)
+        , _permit(std::move(permit))
+        , _as(as)
+        {}
+
+    dht::token consume_end_of_stream() {
+        return _current_key.token();
+    }
+};
+
+view_building_worker::view_building_worker(replica::database& db, view_update_generator& vug, sharded<netw::messaging_service>& messaging) 
+    : _db(db)
+    , _vug(vug)
+    , _messaging(messaging)
+{
+    init_messaging_service();   
+}
+
+future<> view_building_worker::stop() {
+    co_await uninit_messaging_service();
+}
+
+future<> view_building_worker::build_views_range(table_id base_id, dht::token_range range, std::vector<table_id> views) {
+    _as = abort_source();
```

Comment:
This line of code is really surprising. This assigns a new abort_source (side note, I would recommend using `abort_source{}` syntax, I was looking for an `abort_source()` method until I realized this calls the constructor) to a _reference_, so this modifies an outside variable. Why is that?

Reply:
Why `_as` is an outside variable? It's a filed in `view_building_worker` object.

I'm assigning a new abort_source here because it might be aborted before. And once an abort_source was aborted, it cannot be undone.

Reply:
Ok, I got confused. I was looking at the `view_building_worker::consumer` - it has a reference to an abort source. I thought this method here belongs to the consumer.